corpus_studies/solidity_state_study/slither_plugins/statenum.py

```diff
@@ -0,0 +1,332 @@
+from slither.detectors.abstract_detector \
+    import AbstractDetector, DetectorClassification
+from slither.core.expressions import *
+from functools import reduce
+from slither.core.declarations.solidity_variables \
+    import SOLIDITY_VARIABLES,SOLIDITY_VARIABLES_COMPOSED
+from slither.detectors.state.hasstate \
+    import get_vars_used, SOLIDITY_VARIABLE_WHITELIST
+
+# The following functions are convienience functions for tuples.
+def fst(tup) :
+    return tup[0]
+
+def snd(tup) : 
+    return tup[1]
+
+
+# This class is a detector plugin for Slither. It counts the number
+# of states in each contract given to it, and if that number is more than 
+class StateNum(AbstractDetector):
+    """
+    Detect function named backdoor
+    """
+
+
+    ARGUMENT = 'statenum'  
+    HELP = 'Gives the number of states in a given contract'
+    IMPACT = DetectorClassification.HIGH
+    CONFIDENCE = DetectorClassification.HIGH
+
+
+    WIKI = 'STATE COUNT'
+    WIKI_TITLE = 'Number of states'
+    WIKI_DESCRIPTION = 'I been counting states'
+    WIKI_EXPLOIT_SCENARIO = '..'
+    WIKI_RECOMMENDATION = '..'
+    
+    # Takes in two lists and returns one list with the elements of both input
+    # lists, removing duplicates in the process.
+    def list_union(self, list1, list2) :
+        return list(set(list1 + list2))
+
+    # Takes in a list of indeces and a list of upper boundaries on those 
+    # indeces and increments indeces in the index list until an index does
+    # not reach its limit. After that happens, the rest of the input list is
+    # returned unchanged and the indeces that went over their limit are set
+    # to 0. If all elements go over their limits, None is returned instead.
+    def increment_list(self, index_list, length_list) :
+        new_indeces = []
```

Comment:
I think this whole thing might be easier to read in a functional style (with a map).